### ISSUE TITLE
🤯🤯🤯 OMG! Restate + Xstate library

### ISSUE THREAD

**Issue by milanbgd011** [restatedev/xstate#2]
Dear Jack Kleeman, HUGE thank you! This is a major breakthru!!!
Number of SDKs and this Xstate mapping tells me Restate is the real deal.

Anything remotely similar to this, but very basic remapping is here, same thing you built here properly:
https://youtu.be/GpbOkDjpeYU?t=1631

I tried to somehow make the Temporal click with NextJS > therefore and Xstate:
https://github.com/temporalio/samples-typescript/issues/97

I implemented xstate behind the API call that saves snapshot to Postgres, it works but very basic stuff.
With your mapping and power of Restate and Xstate now the possiblities are endless.
Restate Cloud offer seems attractive. 

====

Question:

When will the library be published on NPM so I can try it following the examples you created and try to push it more.

**Comment by milanbgd011** [comment]
Ah, it is published here: https://www.npmjs.com/package/@restatedev/xstate

But you did some updates in the meantime and they are not on NPM. Will this pipeline be placed so that this starts to grow.

I am crusing very difficult tasks with Xstate on the frontend and basic backend, but to be able to do it in a "durable" way is really next level stuff.

**Comment by jackkleeman** [comment]
hi! I am so so glad to see your excitement about this. youre right there are some improvements since 0.0.1, let me do a release and the pipeline will run :)

**Comment by milanbgd011** [comment]
Thank you for pushing the new version.

In this early stage, it would be really valuable to see the full list of tasks that need to be done before the lib reaches version 1.0, just by putting "# Road to 1.0" in README.md and just listing all the tasks that remain to be done.

This will bring clarity and get people involved. I might contribute too if capable to pinch in, but I need a list of clear targets.

You envisioned the solution, it works on the version 0.0.1. When it comes to 1.0 it will move mountains on the shoulders of Restate.

**Comment by jackkleeman** [comment]
What sort of things would you like to see in 1.0? For us, we need to see real world usage before solidifying the featureset and offering a backwards compatibility promise, etc

**Comment by milanbgd011** [comment]
I will definitely try this stuff in production on non-critical stuff to check how it is working.

I would need a bit of time to dig into this deeply and report back. Give me couple of days please and I will post my feedback.

**Comment by kalinicm** [comment]
I got a bit of time, not still digging deep enough unfortunately because of time.

But all works out of the box zero problems:
- invoke
- entry
- exit
- after
- on
- assign

Even remembers the "after" trigger, but when it exists the state where after was initiated then it does not work.

Even if I try to get it again to the state where the "pending AFTER:{} needs to fire" it will reject this "AFTER:{}" because it seems that each state has unique "UUID" which guarantees that the "after" will be executed if that unique ID of the current state is the same, otherwise it will be discarded.

Which means that this "peding" after will never accidentally trigger if we end up in the same state, it simply does not belong to that state unique ID.

```bash
[restate] [auth/send][inv_1h3VdnA6H65X5hvfJibdxkttFWETR00te9][2024-10-19T23:28:02.235Z] INFO:  Received now cancelled event 447f3f3a-61e9-4f54-b8e6-9b1d6bea4994.xstate.after.10000.client.authorizing for target {
  id: 'xyy',
  sessionId: '447f3f3a-61e9-4f54-b8e6-9b1d6bea4994',
  _parent: { id: 'fakeRoot', sessionId: 'fakeRoot' }
}
```

I tried parallel states, works perfectly.

Also tried versioning the machine, like I leave machine in some "invoke" state to wait for 30 seconds, I stop the service, change the machine to go to "stopped" instead of "idle" state and start service. Restate uses new onDone target and just moves to stopped.

====

OMG! What have you built here?! This thing can crush every single task no problem :) 

KUDOS! Not a single problem, glitch or weird behaviour.

This thing goes to production instantly (for non-critical stuff until I am 100% sure no problems will occur).

**Comment by jackkleeman** [comment]
Thank you! What lovely feedback to receive and I am glad you are trying your hardest to break it - definitely keep trying and if you succeed, I will make sure it gets fixed :D 

I think we are missing callback actors, observable actors and possibly transition actors. In practice callback is very similar to promise, observable im not sure really makes sense in a serverless/backend context, but you could certainly make rpcs to the restate ingress from some long-running process with an observable streams. transition actors, i am not sure if these can ever be async - if they are purely sync, they should work as-is, but if async they need a similar approach to promise actors where we run them in a separate handler. definitely more help identifiying particular use cases that are not covered - this would be hugely valuable

**Comment by milanbgd011** [comment]
Breaking it will not be easy task I think :) But sure I will try! Thank you for the offer to fix what I find broken!

Neither of three mentioned actors are imporant. Serverless is very important, so observable is just wrong/weird I think.

> definitely more help identifiying particular use cases that are not covered - this would be hugely valuable

Do you want me to create PR listing all xstate actions, actors... as a bullet points and then we can check off if they work or not in current version. I can confirm many of them and I can test others one by one when I get time, so we have full 1:1 mapping Restate to Xstate and know what is covered and what not. I think visibility of this is key for progress here.

PS. Doing some cool examples and letting David K. Piano like the X post will make this thing visible to all Xstate developers. This is the "marketing channel" perfect to get this off the ground. This is how I found Restate and your repo.

![41108](https://github.com/user-attachments/assets/9cd8187f-4e80-41fb-abc3-b835d51a4fcb)

**Comment by jackkleeman** [comment]
@milanbgd011 that would be fantastic! and any suggestions re examples that perhaps align with what you see as interesting use cases, let me know and I can build them

**Comment by milanbgd011** [comment]
There is no "contribute.md" so I will just post the addtion that could be added to README.md file to track tasks needed to be done before v1.0 is done and also I scanned the entire Xstate V5 docs and created bullet point structure of docs and then gave it a status. Stuff that I tested got ✅, what we talked that will not be supported for now will be ⛔️ and stuff not tested yet are ❓, finally bullet points that are only documentation will have ℹ️. First, this xstate <> restate mapping needs to be tested what works and what does not, then we move to examples and other stuff I suppose.

It would be good to create "CONTRIBUTING.md" and explain how to contribute to the repo properly. Should you enable that anyone can create Merge Request? Now when I push it says I should fork it, so I stopped because there is no CONTRIBUTING.md file.

====>>> please add text below this line to the README.md at the bottom, if you agree on the structure and all

## Road to 1.0

This list will hold all planned things to be integrated before 1.0 is released.

- [ ] users running in production and giving feedback
- [ ] 100% xstate api decision, implementation

## Xstate V5 API tested to be working in Restate

This segment is 1:1 mapping of structure in Xstate V5 docs, then for each option we have status if specific function works inside Restate which uses Xstate machines under the hood. This segment provides clear visiblity what is missing/not implemented and why.

Important to know is that each state has unique ID, so if we schedule "after" to be done in 30 seconds and the machine exists that state and comes back to that state, the unique ID of the current state will be different, so the "after" that we scheadule will not be executed.

- ❓ **Actors**
  - ℹ️ Actors (only documentation)
  - ℹ️ State machine actors
  - ✅ Promise actors
  - ⛔️ Transition Actors (fromTransition()) (not planned to be supported for now)
  - ⛔️ Callback Actors (fromCallback()) (not planned to be supported for now)
  - ⛔️ Observable Actors (fromObservable()) (not planned to be supported for now)
  - ❓ Invoke
    - ℹ️ How are actors different from actions?
    - ℹ️ Lifecycle
    - ℹ️ Re-entering
    - ❓ The invoke property API
      - ✅ Source (fromPromise())
      - ✅ onDone
      - ✅ onError
      - ❓ onSnapshot
      - ❓ Input
    - ✅ Invoking Promises
      - ✅ Promise Rejection
    - ⛔️ Invoking Callbacks (fromCallback()) (not planned to be supported for now)
    - ⛔️ Invoking Observables (fromObservable()) (not planned to be supported for now)
    - ⛔️ Invoking Event Observables (fromEventObservable()) (not planned to be supported for now)
    - ⛔️ Invoking Transitions (fromTransition) (not planned to be supported for now)
    - ✅ Invoking Machines (createMachine())
    - ✅ Sending Responses (sendTo())
    - ❓ Multiple Actors
    - ℹ️ Testing
    - ❓ Referencing Invoked Actors
  - ❓ Spawn
    - ❓ API
    - ❓ Source
    - ❓ Lifecycle
    - ❓ Stopping an actor
  - ❓ Systems
    - ❓ Actor registration
    - ❓ Actor communication
    - ❓ Stopping a system
  - ❓ Inspection
    - ❓ Inspection events
    - ❓ Actor inspection events
    - ❓ Event inspection events
    - ❓ Snapshot inspection events
  
- ❓ **State machine actors**
  - ❓ Promise actors
    - ❓ Promise actor capabilities
    - ❓ Promise actor logic
    - ❓ Promise actor input
    - ❓ Promise actor output
    - ❓ Promise actor error handling
    - ❓ Stopping promise actors
  - ⛔️ Transition Actors (not planned to be supported for now)
  - ⛔️ Callback Actors (not planned to be supported for now)
  - ⛔️ Observable Actors (not planned to be supported for now)
  - ❓ Invoke
    - ℹ️ How are actors different from actions?
    - ℹ️ Lifecycle
    - ℹ️ Re-entering
    - ✅ The invoke property API
      - ✅ Source
      - ✅ onDone
      - ✅ onError
      - ❓ onSnapshot
      - ❓ Input
    - ❓ Invoking Promises
    - ❓ Promise Rejection
    - ⛔️ Invoking Callbacks (not planned to be supported for now)
    - ⛔️ Invoking Observables (not planned to be supported for now)
    - ⛔️ Invoking Event Observables (not planned to be supported for now)
    - ⛔️ Invoking Transitions (not planned to be supported for now)
    - ✅ Invoking Machines
    - ✅ Sending Responses
    - ❓ Multiple Actors
    - ❓ Testing
    - ❓ Referencing Invoked Actors
  - ❓ Spawn
    - ❓ API
    - ❓ Source
    - ❓ Lifecycle
    - ❓ Stopping an actor
  - ❓ Systems
    - ❓ Actor registration
    - ❓ Actor communication
    - ❓ Stopping a system
  - ❓ Inspection
    - ❓ Inspection events
    - ❓ Actor inspection events
    - ❓ Event inspection events
    - ❓ Snapshot inspection events

- ❓ **State machines**
  - ❓ Setup
  - ❓ State
    - ❓ State object
    - ❓ Accessing state snapshots
    - ❓ State value
    - ❓ State context
    - ❓ State children
    - ❓ state.can(eventType)
    - ❓ state.hasTag(tag)
    - ❓ state.matches(stateValue)
    - ❓ state.output
    - ❓ state.getMeta()
    - ❓ State descriptions
  - ❓ Context
    - ❓ Initial context
    - ❓ Lazy initial context
    - ❓ Input
    - ❓ Updating context with assign(...)
  - ❓ Input 
    - ❓ Creating actors with input
    - ❓ Initial event input🔍
    - ❓ Invoking actors with input
    - ❓ Spawning actors with input
    - ❓ Use-cases
    - ❓ Passing new data to an actor
  - ❓ Output
    - ❓ Output in machine actors
    - ❓ Output in promise actors
    - ❓ Output and TypeScript
  - ❓ Events and transitions
    - ❓ Event objects
    - ❓ Selecting transitions
    - ❓ Self-transitions
    - ❓ Transitions between states
    - ❓ Parent to child transitions
    - ❓ Re-entering
    - ❓ Transitions to any state
    - ❓ Forbidden transitions
    - ❓ Wildcard transitions
    - ❓ Partial wildcard transitions
    - ❓ Valid wildcard examples
    - ❓ Invalid wildcard
    - ❓ Multiple transitions in parallel states
    - ❓ Other transitions
    - ❓ Transition descriptions
    - ❓ Shorthands
  - ❓ Eventless (always) transitions
    - ❓ Eventless transitions and guards
    - ❓ Avoid infinite loops
    - ❓ When to use
  - ❓ Delayed (after) transitions
    - ✅ Inlined delays
    - ❓ Referenced delays
    - ❓ Dynamic delays
    - ❓ Lifecycle
  - ❓ Actions
    - ✅ Entry and exit actions
    - ❓ Action objects
    - ❓ Dynamic action parameters
    - ❓ Inline actions
    - ❓ Implementing actions
    - ❓ Built-in actions
    - ✅ Assign action
    - ❓ Raise action
    - ✅ Send-to action
    - ❓ Send-parent action
    - ❓ Enqueue actions
    - ❓ Log action
    - ❓ Cancel action
    - ❓ Stop child action
    - ❓ Modeling
  - ❓ Guards
    - ❓ Multiple guarded transitions
    - ❓ Inline guards
    - ❓ Guard object
    - ❓ Higher-level guards
    - ❓ In-state guards
  - ❓ Initial states
  - ❓ Finite states
    - ✅ Initial state
    - ❓ State nodes
    - ❓ Tags
    - ❓ Meta
    - ❓ Transitions
    - ✅ Targets
    - ❓ Identifying state nodes
    - ❓ Other state types
    - ❓ Modeling states
    - ❓ Finite states and TypeScript
    - ❓ Finite states cheatsheet
  - ❓ Parent states
    - ❓ Root state
    - ✅ Initial state
    - ❓ Transitions on parent states
    - ❓ Child final states
    - ❓ Modeling
  - ✅ Parallel states
  - ✅ Final states
  - ❓ History states
  - ❓ Persistence
    - ✅ Persisting state
    - ✅ Restoring state
    - ❓ Deep persistence
    - ❓ Persisting state machine values
    - ❓ Event sourcing
  - ❓ Tags
  - ❓ Event emitter
    - ❓ Emitting events from actor logic
    - ❓ Emit action creator
    - ❓ Event handlers
    - ❓ Wildcard event handlers

**Comment by milanbgd011** [comment]
@jackkleeman I tested 90%+ of the xstate API I think, all seems to be working, zero issues.

One thing you mentioned on video was not perfectly clear to me, here is video with timestamp where this segment starts https://youtu.be/vECQbpNOQYI?t=1703

QUOTE: `...have some handlers that will always run only one at a time and some handlers that can run concurrently and this is...`
QUESTION: How is this achieved to have handler run only one or to run concurrently? What it actually means?

----

Nice thing is that machine returns the next state machine is in on each CURL sent to it, so that you can use this information to know what is going on the backend immediately, then track the progress with `snapshot` API call.

One thing that could be usefull is to have "event does not exist" when fired, to signify that none of the states/parallel states was able to receive this event and do something with it. But I think this is more xstate territory and must be done manually.

Does the restate remember the events being sent, like a log. Can it be looked into, for example to see which events have been fired and when? I can put it to "log" key in xstate context and log all events in there as they come along with timestamp, but the context would grow rapidly and returning of the snapshot slow because of the size.

**Comment by kalinicm** [comment]
@jackkleeman what I think would be really usefull is to demonstrate how using @resonatedev/xstate library you can tackle this tasks that are on the homepage of the restatedev site. 

This is the examples that would show exactly how these scenarios are handled in xstate solution.

<img width="1454" alt="Screenshot 2024-10-23 at 09 01 01" src="https://github.com/user-attachments/assets/2ae6e941-ee6f-4750-a74b-8be9b861cfba">

**Comment by jackkleeman** [comment]
> QUOTE: ...have some handlers that will always run only one at a time and some handlers that can run concurrently and this is...
> QUESTION: How is this achieved to have handler run only one or to run concurrently? What it actually means?

In Restate, some types of services are 'keyed' and they can store state. For example, virtual object services have this property, and the xstate library is built on top of a virtual object. For a given key, handlers of a virtual object will by default lock the object and can not run concurrently with each other. However, you can also set a handler to be 'shared' which means it can run concurrently with any other handler running, but it can't mutate state, and it only reads a snapshot of state. In the xstate integration, the invokePromise handler is one such handler, which doesn't block the main loop. This is necessary because it can run for arbitrarily long to complete a promise execution. However, all other handlers do not block (in line with xstate API) and so should run for very short periods of time. These are executed serially, so that we have strong ordering, and so that they can safely update state instead of just read it.

---

> One thing that could be usefull is to have "event does not exist" when fired, to signify that none of the states/parallel states was able to receive this event and do something with it. But I think this is more xstate territory and must be done manually.

Agreed. I think it is possible to interpret the state machine at runtime to determine the set of valid events for the current state, and reject invalid events. If you file a separate issue I can take a look at this!


---

Re recurring events, i think broadly the idea of repeatedly polling via self-transitions (after: 1h -> self) makes sense. I think you can use child machines to have multiple of these polls concurrently executing. I would hope the approach to this is the same with our integration as it would be in native xstate.
As for .sleep() - this is a way to suspend a handler invocation for an arbitrarily long time. The actual `after` feature is implemented with a delayed invocation eg `.send("hello", {delay: 1000})`. This is because in xstate we don't want to block an invocation as this will block the whole state machine, we instead want to schedule non-blocking events in the future. In normal restate code without xstate, you might use a mix of these two approaches depending on the use case.

**Comment by milanbgd011** [comment]
@jackkleeman Thank you for the detailed response. 

1.
I will express how I percieve the information you shared, then you correct me if I am wrong.

BLOCKING (should be instant, very fast to execute because they block machine)

- assign()
- enqueue.assign()

NON-BLOCKING (does not block machine, we can send more events to it no problem)

- raise() - execute now (+ delay)
- enqueue.raise() - execute after entering fully next state (after entry:{}) (+ delay)
- sendTo() (+ delay)
- enqueue.sendTo() (+ delay)
- sendParent() (+ delay)
- invoke: fromPromise()
- target: '...'

2.
I think that I fired event against machine with 2 parallel states, while one state was inside "invoke: fromPromise() waiting 5 seconds to resolve". It did take the new event and started transition on the second parallel state without affecting current processing of first state. Is this expected? This is perfect scenario basically. Machine responds to events if possible to receive them on matter which parallel state is doing what.

3.
When we are doing the "blocking" stuff, then all incoming events are being scheduled while machine is "stuck" doing some assign(), but none of the events will drop.

**Comment by jackkleeman** [comment]
1. I am not sure that any in-xstate operation can block the state machine. From Restates perspective, every invocation through this library is just read state -> do something which will not include any `await` -> write state -> finish. as such, invocations should always be very short. with the sole exception of invokePromise handler, where promises are executed, and there is an `await`, but this handler doesnt block the machine. 

2. it is 100% expected that the machine can still process events while some part of it is waiting on a fromPromise execution. This is because the promise is offloaded to the `invokePromise` handler which is a restate shared handler, and so can execute concurrently to other handlers. All the handler needs to do is find the promise actor definition in the state machine, pass the appropriate input values, await the promise, and send the result or rejection back to the main state machine

3. It is not possible to do any blocking stuff within this library, because we simply do not `await` anything in any handler except `invokePromise`. Even using Restates underlying features available on the restate.ObjectContext, I don't think you can block the state machine without that `await`. However, invocations of course still take time to complete (a few milliseconds), during which incoming events queue up inside Restate. Certainly no events will ever drop. If you did somehow manage to block the state machine meaningfully, for example if you took down to the lambda/node service running the underlying code such  that the invocation cannot be processed, what this would look like is a long invocation on a handler like `send`, ie an event is taking a long time to be processed. To make an analogy, if you were using xstate in the browser, this would be like the browser hanging, during which time xstate can't receive any new messages from the event loop

**Comment by milanbgd011** [comment]
@jackkleeman then you for the clarification, it seems very simplistic to reason about, just like a simple xstate machine - but a durable execution one.

I am not sure how you built this resilient thing in couple of commits really. Seems really weird :)

I guess the only thing for me to do is just to try to break it and report back.

From the examples perspective, first one needed is with extreme heavy usage of scheaduled events (or anything related S to time) could be a good candidate, because simple execution is ok, but doing this durably. Demoing how you can do newsletters, subscriptions and other time related stuff.

Really exciting about this xstate + restate. Maybe it is not obvious for others how powerful this combinaiton, but sure enough they will know. Thank you once again for the effort, really appreciated.

Keeping you posted.

**Comment by jackkleeman** [comment]
> I am not sure how you built this resilient thing in couple of commits really.

I guess the xstate bit is quite simple, but the restate bit is a lot of commits ;)

Awesome ideas!

Re machine versioning, this is a hard problem in general in durable execution. In some ways its a bit more tractable using XState, because you only have to ensure that the machine state is still valid under the new version, ie the states, inputs etc for the *current state* (of all existing state machines) still make sense in the new version. Currently to solve the more general versioning problem users can inform Restate of new versions and it will send new invocations to the new versions; however in the case of this library, the new version is still going to read the state set by the old version, so there is still a need for state-compatibility. 

I think the safest way to do versioning when making changes that can't be state-compatible is to have multiple state machine definitions side by side inside a parent state machine. The first step of the parent state machine would be to select the latest version available in the running service, then delegate to that machine. This selected version is then persisted by xstate like any other state node, and if later new versions are added, we will keep using the old state machine, but new invocations will use the new one. I would hope this is only needed in exceptional circumstances however

**Comment by milanbgd011** [comment]
Rule is simple, do not delete old state, just rewire it to forward to right place and keep it in code until you are sure 100% no machines running depend on it. It does not bother if you have these old states lingering around, it is ok. It is way better than doing any versioning which is very hard to maintain.

We have 2 different types of "versioning":

- versioning existing EVENT
  - for the event it is easy, just update the existing event listener to do something else 
- versioning existing STATE
  - in old state listen to all events, target new state and enqueue event that is sent to the new state


### Versioning the state example

V1

```js
{
     idle: {},
     awaitingConfirmation: {
       on: {
         CONFIRM: {...},
         .....more events
       },    
     },
}
```

V2

It will react to any event sent to machine and forward this event to new state the machine should be in

```js
{
     idle: {},
     awaitingConfirmation2: {
        on: {
         CONFIRM: {...},
         .....more events
        }
     },
    // inactive states (handle old machine versions)
     awaitingConfirmation: {
       on: {
         '*': {
            target: 'awaitConfirmation2',
             actions: [
                 ({ enqueue, event }) => {
                      enqueue(event);
                 }
             ]
          }
       }
     },
}
```


actions:

**Comment by jackkleeman** [comment]
Your examples re versioning makes sense to me

I agree it would be cool to write a post comparing kafka/temporal to resstate/xstate. Its definitely on our to do list!

I think Restate can definitely handle a large volume of updates, Kafka style. Restate is internally a streaming system and we've benchmarked thousands of events per second on a single machine - and we can horizontally scale, too. I don't think you would need to bring in another tool like Redis streams. That said, you may not want to use the xstate integration for that kind of volume - for maximum throughput you might want a Go or Rust service, using Restates default RPC API, instead of the Xstate one.

**Comment by milanbgd011** [comment]
Thank you Jack!

In the plethora of tools and approaches, it is easy to miss the stuff Restate brings to the table actually.

Creating dummy "uber" app using restate/xstate (with RPC for high thruput GPS location and xstate for all other) and then comparing to Temporal/Kafka, and pinpointing the "weak spots" in their stack will pull in people to give it a try. Building good starting point for others to experiment seems legit. 

So basically, "let us wire up our product to simplistic but complete real world scenario and then show off where they shine". 

You know, even all cars seem alike but deeply they are not, but this is very hard to spot without spending a ton of time and investing yourself to learn new framework/tool before going to prod. So, many will not do it unfortunately.

Restate goes straight to the backbone of the entire company, whoever decides to use it. Very specific situation. :)

I have perfect project for restate/xstate duo, so I hope to get my hands dirty soon.

**Comment by milanbgd011** [comment]
> we can horizontally scale, too.

Since restate uses "internal database" and RocksDB for some additional stuff.
Question is how do you actually scale this horizontally since each replica is StatefulSet (with own database).
Cloud offering seems very attractive when the business grows and reliability is crucial, but until one project reaches this level question is just how to have 2 pods in kubernetes working as horizintal scaling to be able to have at least basic levels of reliability for starters.

I could not find this information at all. Replication is very critical part of the horizontally scaled Restate server. I mean, when you look at the levels TigerBeetleDB reaches on this stuff then you appreciate the complexity of reliable replication more: https://sim.tigerbeetle.com/

**Comment by jackkleeman** [comment]
Its not released yet (should have something by end of year), but Restate is implementing a replicated distributed log similar to delos (https://www.usenix.org/system/files/osdi20-balakrishnan.pdf). The log is indeed sharded on partition key which is derived from the virtual object or workflow key (or is random for nonkeyed services). This means that you can scale horizontally in the number of keys very easily by adding more servers.

Restate IS a database and our goal for doing this is not one of business defensibility but one of performance. If you use a general purpose database for this, you get all their trade offs that are intended to be sensible for a broad range of use cases. By building our own storage layer, every little trade off decision in the DB is targeted to this quite unique use case. the internals of restate are much closer to something like kafka than they are to something like postgres, so really we are more of a broker than a DB

**Comment by milanbgd011** [comment]
@jackkleeman Restate just keeps on giving. Sorry for the hard questions, but this last comment clarifies the idea behind Restate much deeper for me personally. Will read the PDF soonish.

Each performance squeezed out is huge influence to the cloud offering and growth of the project. Being able to delegate this responsibility to the creators of the project and just relax - no price on that. Since it is so critical system, stable cloud will deliver so much value to any business. Even allow me to create a bunch of projects quick and not think about infra at all.

Looking forward having this "superpower" for building really complex apps that generate huge value for end users.

Thank you!
